.werft/build.ts

```diff
@@ -289,6 +314,254 @@ interface DeploymentConfig {
+
+    // find free ports
+    werft.log(installerSlices.FIND_FREE_HOST_PORTS, "Check for some free ports.");
+    const [wsdaemonPortMeta, registryNodePortMeta, nodeExporterPort] = findFreeHostPorts([
+        { start: 10000, end: 11000 },
+        { start: 30000, end: 31000 },
+        { start: 31001, end: 32000 },
+    ], metaEnv({ slice: installerSlices.FIND_FREE_HOST_PORTS, silent: true }));
+    werft.log(installerSlices.FIND_FREE_HOST_PORTS,
+        `wsdaemonPortMeta: ${wsdaemonPortMeta}, registryNodePortMeta: ${registryNodePortMeta}, and nodeExporterPort ${nodeExporterPort}.`);
+    werft.done(installerSlices.FIND_FREE_HOST_PORTS);
+
+    // clean environment state
+    try {
+        if (deploymentConfig.cleanSlateDeployment) {
+            werft.log(installerSlices.CLEAN_ENV_STATE, "Clean the preview environment slate...");
+            // re-create namespace
+            await cleanStateEnv(metaEnv());
+
+        } else {
+            werft.log(installerSlices.CLEAN_ENV_STATE, "Clean the preview environment slate...");
+            createNamespace(namespace, metaEnv({ slice: installerSlices.CLEAN_ENV_STATE }));
+        }
+        werft.done(installerSlices.CLEAN_ENV_STATE);
+    } catch (err) {
+        werft.fail(installerSlices.CLEAN_ENV_STATE, err);
+    }
+
+    // Now we want to execute further kubectl operations only in the created namespace
+    setKubectlContextNamespace(namespace, metaEnv({ slice: installerSlices.SET_CONTEXT }));
+
+    // trigger certificate issuing
+    try {
+        werft.log(installerSlices.ISSUE_CERTIFICATES, "organizing a certificate for the preview environment...");
+
+        // trigger certificate issuing
+        await issueMetaCerts(namespace, domain);
+        await installMetaCertificates(namespace);
+        werft.done(installerSlices.ISSUE_CERTIFICATES);
+    } catch (err) {
+        werft.fail(installerSlices.ISSUE_CERTIFICATES, err);
+    }
+
+    // add the image pull secret to the namespcae if it doesn't exist
+    const hasPullSecret = (exec(`kubectl get secret ${IMAGE_PULL_SECRET_NAME} -n ${namespace}`, {slice: installerSlices.IMAGE_PULL_SECRET, dontCheckRc: true, silent: true })).code === 0;
+    if (!hasPullSecret) {
+        try {
+            werft.log(installerSlices.IMAGE_PULL_SECRET, "Adding the image pull secret to the namespace");
+            const auth = exec(`echo -n "_json_key:$(kubectl get secret ${IMAGE_PULL_SECRET_NAME} --namespace=keys -o yaml \
+                | yq r - data['.dockerconfigjson'] \
+                | base64 -d)" | base64 -w 0`, { silent: true }).stdout.trim();
+            fs.writeFileSync(`./${IMAGE_PULL_SECRET_NAME}`,
+            `{
+    "auths": {
+        "eu.gcr.io": {
+            "auth": "${auth}"
+        }
+    }
+}`);
+            exec(`kubectl create secret docker-registry ${IMAGE_PULL_SECRET_NAME} -n ${namespace} --from-file=.dockerconfigjson=./${IMAGE_PULL_SECRET_NAME}`);
```

Comment:
What's wrong wiht the format of IMAGE_PULL_SECRET_NAME so that you need to extract the "auth" from it only to package it into another secret? :)

Reply:
Great question! 

I studied the secret that gets created for preview environments when built using Helm. If you compare one of those, with the one in `keys` namespace, you'll see they do not match in type or object structure too.

The secret in `keys` is opaque, I assume its purpose is to be a single source of truth. For example, each time a preview environment is created, get it, and then use its contents to build a new docker registry secret for the namespace being created for the preview environment.

For example:
```
gitpod /workspace/gitpod $ kubectl get secret gcp-sa-registry-auth -n keys
NAME                   TYPE     DATA   AGE
gcp-sa-registry-auth   Opaque   1      127d
gitpod /workspace/gitpod $ kubectl get secret gcp-sa-registry-auth
NAME                   TYPE                             DATA   AGE
gcp-sa-registry-auth   kubernetes.io/dockerconfigjson   1      4m45s
```
I believe the `auths` structure is specific to GCR.

Let me know if you need anything else?

Reply:
Exactly @kylos101. In the past we used the same info to build different types of secrets out of it, that's why it's not a simple copy. Not sure if that's the case anymore.